src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,148 @@
+            head.prev = newNode;
+            head = newNode;
+            size++;
+        } else {
+            Node<T> current = getNodeByIndex(index);
+            newNode = new Node<>(current.prev, value, current);
+            current.prev.next = newNode;
+            current.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, size);
+        Node<T> oldValue = getNodeByIndex(index);
+        T oldItem = oldValue.item;
+        oldValue.item = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size);
+        Node<T> getNode = getNodeByIndex(index);
+        T removeElement = getNode.item;
+        unLink(getNode);
+        return removeElement;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.item == object || (current.item != null) && current.item.equals(object)) {
+                unLink(getNodeByIndex(i));
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index, int size) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Wrong index");
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkIndex(index, size);
+        Node<T> current = head;
+        if ((size / 2) > index) {
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private void unLink(Node<T> removedElement) {
+        if (isEmpty()) {
+            throw new RuntimeException("Can't remove it");
+        } else if (size == 1) {
+            head = null;
+            tail = null;
+        } else if (removedElement == head) {
+            head = removedElement.next;
+            head.prev = null;
+        } else if (removedElement == tail) {
+            tail = removedElement.prev;
+            tail.next = null;
+        } else {
+            removedElement.prev.next = removedElement.next;
+            removedElement.next.prev = removedElement.prev;
+        }
```

Comment:
Let's refactor this code to encrease readability

```
        Node<T> prev = removedElement.prev;
        Node<T> next = removedElement.next;

        if (prev == null) {
            head = next;
        } else {
            prev.next = next;
            removedElement.prev = null;
        }
```
The similar check for `(next == null)`.
And set  `inputNode.item = null` in the end.
Where `Node<T> inputNode` - is method input parameter.